TFEngine.py

```diff
@@ -1344,48 +1345,81 @@ def _do_save(self):
         return False
     return True
 
-  def eval_model(self, output_file=None, output_per_seq_file=None):
+  def eval_model(self, output_file=None, output_per_seq_file=None, loss_name=None, output_per_seq_format=["seq_len", "score", "pos_score"]):
     """
     Eval the current model on the eval datasets (dev + eval, whatever is set).
     See also :func:`self.search` for performing beam search.
 
     :param str|None output_file: if given, will save the results to this file (total err/score for each dataset)
     :param str|None output_per_seq_file: if given, will save the err/score for each sequence
+    :param str|None loss_name: specifies the loss which will be written to output_file
+    :param list[str]|tuple[str]|None output_per_seq_format:
+      which properies of `loss_name` should be written to `output_per_seq_file`.
+      allowed_outputs = {"seq_tags", "seq_len", "score", "error", "pos_score", "pos_error"}
+
     :return: nothing
     """
+
     extra_fetches = None
-    results_per_seq = {}  # seq_tag -> dict[str,float]
+
     if output_per_seq_file:
-      extra_fetches = {}
+      allowed_outputs = {"seq_tags", "seq_len", "score", "error", "pos_score", "pos_error"}
+
+      assert isinstance(output_per_seq_format, (tuple, list))
+      assert set(output_per_seq_format) - allowed_outputs == set(), \
+        "Only %s are allowed in function eval_model as output_per_seq_format, but got: %s " % (allowed_outputs, output_per_seq_format)
+
+      # always fetch seq_tag to map loss values to the corresponding line
+      extra_fetches = {"seq_tags": self.network.get_seq_tags()}
+
       from TFUtil import identity
       losses_dict, _, _ = self.network.get_losses_initialized(reduce_func=identity, with_total=False)
-      for loss_name, loss_holder in losses_dict.items():
-        loss_holder.loss.layer = None  # enforce reinit
-        loss_holder.loss.use_flatten_frames = False  # we need that such that we get (B*T,...) unreduced values
-        loss_value = loss_holder.get_normalized_loss_value_per_seq()
-        error_value = loss_holder.get_normalized_error_value_per_seq()
-        if loss_value is not None:
-          extra_fetches["loss:score:%s" % loss_name] = loss_value
-        if error_value is not None:
-          extra_fetches["loss:error:%s" % loss_name] = error_value
-        if loss_holder.loss.output.have_time_axis():
-          seq_lens = loss_holder.loss.output.get_sequence_lengths()
-          extra_fetches["loss:seq_lens:%s" % loss_name] = seq_lens
-        extra_fetches["seq_tags"] = self.network.get_seq_tags()
-      assert extra_fetches
-
+      assert loss_name in losses_dict, \
+        "Unknown loss defined. Got \"%s\". Possible losses are %s" % (loss_name, losses_dict.keys())
```

Comment:
Use `%r` instead of `\"%s\"` and also instead of the second `%s`.
Also, better use brackets instead of `\` to break the newline, as in `assert loss_name in losses_dict, (\n "Unknown loss ...")`, as this makes it easier to identify in what line the assertion happened, when the code is parsed, for tools like better_exchook. Which is some useful information here, as the user would basically see also the variable names, which can be helpful.